_posts/2020-09-16-hibernate-search.md

```diff
@@ -0,0 +1,440 @@
+  QueryBuilder qb = fullTextEntityManager
+            .getSearchFactory()
+            .buildQueryBuilder()
+            .forEntity(User.class)
+            .get();
+  
+  Query similarToUser = qb.keyword().fuzzy()
+            .withEditDistanceUpTo(2)
+            .onField("first")
+            .matching(first)
+            .createQuery();
+  Query finalQuery = qb.bool()
+            .must(similarToUser)
+            .createQuery();
+  FullTextQuery fullTextQuery = 
+            fullTextEntityManager.createFullTextQuery(finalQuery, User.class);
+  fullTextQuery.setProjection(FullTextQuery.ID, "first","last","middle","age");
+  fullTextQuery.setSort(qb.sort()
+                            .byField("age")
+                            .desc()
+                            .andByScore()
+                            .createSort());
+  fullTextQuery.setMaxResults(max);
+  fullTextQuery.setFirstResult(page);
+  return getUserList(fullTextQuery.getResultList());
+}
+
+private List<User> getUserList(List<Object[]> resultList) {
+  List<User> users = new ArrayList<>();
+  for (Object[] objects : resultList) {
+      User user = new User();
+      user.setId((String) objects[0]);
+      user.setFirst((String) objects[1]);
+      user.setLast((String) objects[2]);
+      user.setMiddle((String) objects[3]);
+      user.setAge((Integer) objects[4]);
+      users.add(user);
+  }
+  return users;
+}
+```
+
+Before moving on to the main event let's zoom in on a couple of interesting things here.
+
+* Up until now we were using keyword queries to perform exact match searches only, but here as you can see I have combined
+it with the fuzzy query which will allow us to perform fuzzy searches too. This query gives end-users some flexibility in terms of searching by allowing some degree of error. The threshold of the error to be allowed can be decided by us. For instance, here I have
+set edit distance to 2 (Default is also 2 by the way) which means Elasticsearch will match all the words with a maximum of 2 differences to 
+the input. e.g., 'jab' will match 'jane'
+
+* We can also perform sorting and pagination. Sorting can be applied on single or multiple fields, and pagination can be 
+controlled with `setMaxResult`--Number of results we need per page-- and `setFirstResult`--We can pass page number in it.
+
+* Finally, let's talk about projection. To use projection we need to pass the list of fields that we want in output in the `setProjection` method. 
+Now when we fetch results hibernate will return a list of object arrays. Which to be honest is a bit incontinent. Apart from
+fields, we can also fetch metadata such as id with `FullTextQuery.ID` or even score with `FullTextQuery.SCORE`.
+
+That's it! I mean this is not everything, but I believe this is enough to get you started. For further reading you can 
+explore the following:
+
+* Phrase queries - which allows you to search complete sentences
```

Comment:
Would be nice if you added links in this list, that makes a "Further Reading" section so much more valuable.